### PR TITLE
fix(rag): distill the retrieval query so the lexical side actually fires

### LINKED ISSUE #198: [P3][High] Retrieval query is the raw JD: the lexical side never fires and the dense query is truncated
Part of #152 (Phase 3 — make the flagship AI claims true).

## The bug

`retrieve_resume_evidence` (`app/rag/store.py:203`) passes the **entire job description** as the
retrieval query. Two things break:

**1. The lexical side is structurally dead.** `_lexical_candidates` (`store.py:105`) runs
`websearch_to_tsquery('english', <whole JD>)`, and `websearch_to_tsquery` **ANDs** bare terms.
A 400-word JD becomes a conjunction of ~250 non-stopword lexemes that no résumé chunk can
satisfy, so the FTS side returns **zero rows**. RRF then fuses `[dense, []]` — which is just
`dense`. **`hybrid` is mathematically identical to `vector`**, and the eval's "hybrid ≈ vector,
within judge variance" conclusion is not a finding about hybrid retrieval at all.

Note this is *not* the failure `_fts_ready()` guards against — the column and index exist, so the
mode is reported as measured rather than N/A.

**2. The dense side is silently truncated.** `all-MiniLM-L6-v2` has a 256-word-piece limit
(`app/rag/embeddings.py:11`). A full JD is truncated to roughly its first paragraph — usually
company boilerplate — so the dense query vector represents the header, not the requirements.
The cross-encoder reranker gets the same over-long query.

## The fix

Distill a retrieval query from the JD instead of using it raw. `parse_job` already extracts
`required_skills` / `preferred_skills` / `title` — that is exactly the signal retrieval wants,
and `ScoreFitRequest` already carries `required_skills` (`app/schemas.py:115`).

- [ ] Derive the retrieval query from parsed skills + title, falling back to a bounded
      keyword extraction when no parse is available.
- [ ] Keep the query inside the embedding model's token budget; log/trace the distilled query.
- [ ] Consider `OR` semantics for the lexical side so partial matches rank rather than
      requiring every term.
- [ ] Add a test that fails when the lexical candidate set is empty for a realistic JD —
      the missing test that let this ship.
- [ ] Re-sweep on top of #197's honest harness and re-state whether hybrid/rerank actually
      beat vector now that the lexical side can fire.

## Acceptance

For a realistic JD + résumé pair, the lexical side returns a non-empty candidate set, `hybrid`
produces a ranking that differs from `vector`, and the published comparison reflects a real
hybrid measurement.

Depends on #197 (the harness must be honest before any re-measurement means anything).

### PR DESCRIPTION
**Pull request by Taleef7**
Closes #198. Second PR of Phase 3 (epic #152).

## The bug

`retrieve_resume_evidence` passed the **whole job description** as the retrieval query.
`websearch_to_tsquery` **ANDs** bare terms, so a JD compiled to a ~98-node conjunction no résumé
chunk could satisfy. The lexical side matched **0 of 16** gold-set JDs, RRF fused `[dense, []]`
(= `dense`), and "hybrid" was **byte-identical to vector on 16/16 rows**.

Every hybrid-vs-vector conclusion this project ever published was therefore a conclusion about
vector.

## The fix

`app/rag/query.py` distills a JD into **title + parsed skills** — `parse_job` already extracts
exactly that signal, and the API parses before scoring, so they're populated in production. The
lexical terms are then **quoted and OR-joined** so `ts_rank_cd` can rank partial matches.

Two parser hazards found by probing Postgres directly rather than trusting the docs, both now
pinned by tests:

| input | parses as | consequence |
| --- | --- | --- |
| `-drop` (unquoted) | `!!'drop'` | negation — *excludes* the chunks we want |
| `a"b` | `'b' & 'python'` | quote closes early, silently restoring **AND** |

The keyword fallback also needed real work. Plain frequency ranking returned
`Acme, partners, Corp, where, can` and missed every technology — JDs repeat boilerplate in the
header and name their requirements once at the end. It now mines the **requirements section**.

## But fixing the query wasn't enough — the eval couldn't measure retrieval at all

The gold résumé chunked into **exactly 4 pieces** and the sweep retrieves **k=4**. Every
"retrieval mode" returned the whole résumé, reordered. That invalidates a claim merged
yesterday in #197 — *"top-k retrieval recovers full-résumé quality from a fraction of the
context"* — because top-4-of-4 is not a fraction.

So `sample_resume.txt` is expanded 1.5 KB → 4.2 KB (**4 → 9 chunks**), making `k=4` a real ~45%
selection. **The qualification profile is deliberately unchanged**: 10 of the 16 gold rationales
turn on technologies being *absent* ("Django/Flask not in the resume", "Java, Spring Boot,
Kubernetes and Kafka are absent"), so only detail was added — same skills, same ~3 years, same
seniority. A word-boundary check confirms no disqualifying term appears and every positive-match
skill survives. **Worth your eyes, since the labels are yours.**

## Results (gpt-4o-mini judge, 9-chunk résumé, k=4, n=16, 0 errors)

| mode | Spearman | faithfulness | ans. rel. | ctx recall |
| --- | --- | --- | --- | --- |
| off (résumé-blind) | 0.268 | 0.387 | 0.369 | 0.333 |
| vector | 0.726 | 0.769 | 0.293 | 0.427 |
| hybrid | **0.757** | 0.729 | 0.230 | 0.479 |
| hybrid+rerank | 0.729 | 0.704 | 0.180 | 0.417 |
| full-resume | 0.586 | 0.786 | 0.163 | **0.521** |
| full-resume+vector | 0.594 | **0.807** | 0.206 | 0.490 |

**The lexical side is alive:**

| | before | after |
| --- | --- | --- |
| JDs whose lexical query matches ≥1 chunk | 0/16 | **16/16** |
| rows where hybrid retrieves different chunks than vector | 0/16 | **13/16** |

### The genuinely interesting result

**Retrieval outranks the whole résumé.** `vector` 0.726 vs `full-resume` 0.586 — a **0.140 gap,
2.2× the noise floor**, with `full-resume` below the *entire* 5-replicate range of `vector`
(0.771–0.835). Feeding all nine chunks makes the model rank candidates **worse** than feeding it
the four retrieval picked.

That inverts the usual framing. Retrieval here isn't a lossy compromise accepted for
context-window reasons — it's a **precision filter that improves the judgment**. (Direction is
ranking-specific: faithfulness slightly favours the full résumé, well inside noise.)

### What is still unresolved

`hybrid` vs `vector` is Δ0.031 — half the floor. The reranker is Δ0.003. Both are honest nulls
now rather than impossibilities. A 16-row set with one résumé is a weak instrument for detecting
a re-ranking win.

## Noise floor is corpus-specific — re-measured, not inherited

| | old corpus | new corpus |
| --- | --- | --- |
| Spearman max pairwise Δ | 0.076 | **0.063** |
| faithfulness max pairwise Δ | 0.080 | **0.120** |

Faithfulness noise grew by half. Inheriting the old threshold would have mis-graded results in
both directions. And **for the second time a single sweep value landed outside five replicates
of the same configuration** (vector Spearman 0.726 vs a 0.771–0.835 range) — five replicates
estimate the spread, they don't bound it.

Also retracted: the context-recall "judge sanity check" from #197. `full-resume`'s higher recall
(Δ0.094) is inside the 0.125 recall noise band, so it can't serve as one.

## Verification

- **210 agent tests green**, 5 skipped (real-DB, no local `DATABASE_URL`), ruff clean
- 14 unit tests + **5 against a real Postgres**, wired into CI's existing pgvector `db` job —
  psycopg only, no embeddings/torch. Mocked cursors happily accept a tsquery matching nothing,
  which is exactly how this shipped; the real-DB test pins both the old (empty) and new
  (non-empty) behaviour
- Sweep and noise runs executed against the live DB; both reports committed as artifacts

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: EVALS.md [hunk 1]
```diff
@@ -90,38 +90,57 @@ Fixed in [#197](https://github.com/Taleef7/jobops-copilot/issues/197): the judge
...
-On this gold set **`hybrid` and `vector` are the same experiment.** The lexical side is
-structurally dead: `_lexical_candidates` builds `websearch_to_tsquery('english', <whole JD>)`,
-which **ANDs** every term, so a JD becomes a ~98-node conjunction that no resume chunk can
-satisfy. Verified directly against the database: **0 of 16** JDs match a single chunk, and
-`hybrid` returns **byte-identical chunks to `vector` on 16/16 rows**. RRF fusing
-`[dense, []]` is just `dense`.
+Before #198 the FTS half of "hybrid" was dead: `websearch_to_tsquery('english', <whole JD>)`
+**ANDs** every term, so a JD became a ~98-node conjunction no chunk could satisfy. RRF fusing
+`[dense, []]` is just `dense`, so hybrid was byte-identical to vector. Measured against the
+database, before and after distilling the query to title + parsed skills, OR-joined:
```
[the remark this answers is not in the document]
Comment (Taleef7) [reply EVALS.md:141]: Correct, and it changed the result. Fixed in 3519e55.

Confirmed exactly as described: `evals/retrieval.py` passed only the résumé and raw JD, so `terms_for()` took its keyword-fallback path while `/score-fit` retrieves with parsed skills. The committed table described a top-k selection the product never performs — the same class of error #197 existed to fix, one layer down.

**Fix.** The sweep now joins each fit-score row to the hand-labeled `parse_job.jsonl` gold by id and passes `title` + `required_skills`. **The noise harness had the identical gap** and got the same treatment, so the spread is measured on the configuration actually being compared.

I used the *gold* parse rather than calling `parse_job` live, deliberately: a live parse would inject extraction variance into a comparison whose deltas already sit near the noise floor. 14 of 16 rows have a gold parse; the other two fall back to keywords, and that path is asserted by test. Queries now look like `['Python Developer', 'Python', 'REST APIs', 'Django', 'Flask', 'SQL', 'Git']`.

**Re-running flipped a headline.** Since a between-mode claim needs *both* modes replicated — a single sweep value has now escaped its own five-replicate range three separate times — `--noise-floor` takes a `mode`, and I replicated hybrid as well as vector:

| configuration | Spearman mean | range | n |
| --- | --- | --- | --- |
| `vector` | 0.716 | 0.706 – 0.733 | 5 |
| `hybrid` | **0.821** | **0.800 – 0.848** | 5 |

**Hybrid beats dense-only.** The ranges don't overlap — hybrid's worst run exceeds vector's best by 0.067, mean difference 0.105, Welch's t ≈ 9.9. That's the first retrieval improvement this project has been able to demonstrate, and it only became measurable once this PR revived the lexical side: before, hybrid retrieved byte-identical chunks and *could not* differ.

The single sweep understated it (Δ0.088) because that run's `vector` was a high draw above its own replicate range.

The gain is **ranking-specific** — faithfulness moves the other way (hybrid 0.713 vs vector 0.748), unresolved against hybrid's 0.103 faithfulness spread. Reported as such.

Also worth recording: noise floors are configuration-specific and keep moving — Spearman 0.076 → 0.063 → **0.027**, faithfulness 0.080 → 0.120 → **0.040** across the three gold-set/query combinations measured. A focused deterministic query yields markedly more consistent generations. Never inherit a threshold.

File: services/agent/app/main.py [hunk 2]
```diff
@@ -257,7 +257,13 @@ def score_fit_endpoint(req: ScoreFitRequest) -> FitScoreResponse:
...
         if rag_available() and req.resume_text and not req.retrieved_context:
             evidence = retrieve_resume_evidence(
-                req.resume_text, req.description_text, user_id=req.user_id
+                req.resume_text,
+                req.description_text,
+                user_id=req.user_id,
+                # The API parses the job before scoring, so these are normally
+                # populated; they are what the distilled query is built from (#198).
+                required_skills=req.required_skills,
+                preferred_skills=req.preferred_skills,
```
[the remark this answers is not in the document]
Comment (Taleef7) [reply services/agent/app/main.py:266]: Confirmed and fixed in 3519e55.

`ScoreFitRequest` had no `title` field, so the title produced by the parse immediately upstream was dropped before it could reach `terms_for()`. And because any parsed skill suppresses the keyword fallback, your example is exactly right: a role whose only parsed skill is generic ("communication") retrieved on that word alone, with nothing identifying the role.

Wired end to end:

- `app/schemas.py` — `ScoreFitRequest.title`, documented with the reason it exists
- `app/main.py` — passes `title=req.title` into `retrieve_resume_evidence`
- `apps/api/src/lib/agent-client.ts` — `ScoreFitInput.title`, sent as `title` in the `/score-fit` body
- `apps/api/src/routes/ai.ts` and `routes/n8n.ts` — both call sites pass `parsed.title`

`ParsedJobOutput.title` already existed at both call sites, so nothing new had to be plumbed through the parse step.

Verified in the distilled output — the title now leads the query:

```
adzuna-8   -> ['Python Developer', 'Python', 'REST APIs', 'Django', 'Flask', 'SQL', 'Git']
adzuna-33  -> ['AI / LLM Engineer', 'Generative AI', 'Machine Learning', ...]
adzuna-35  -> ['Gen AI Developer / Engineer', 'Python', 'LLMs', 'NLP', ...]
```

The eval sweep passes the gold title through the same path (see the P1 thread), so the measured retrieval and the production retrieval now agree on this field too.

192 API tests and `tsc --noEmit` clean.
